### PR TITLE
Avoid calling queries during query stack printing

### PR DESCRIPTION
**Pull request by flip1995**
This has the side effect, that when Clippy should ICE (during an EarlyPass?) it will fill up the RAM with 2 GB/s and then freezes my Laptop. This is blocking the Clippy sync and might give some people really bad experiences, so this should be merged ASAP.

r? @cjgillot
cc @Zoxc 

I only commented this on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.60try_print_query_stack.60.20has.20.60ImplicitCtx.60.20during.20.60EarlyPass.60/near/363926180). I should've left a comment on the PR as well. My bad.

### CODE REVIEW

**Comment by cjgillot** [comment]
@flip1995 do you have a reproducer to add as a test? We wouldn't want this to come back again.
@Zoxc any idea of what's happening?

**Comment by flip1995** [comment]
I posted a reproducer on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.60try_print_query_stack.60.20has.20.60ImplicitCtx.60.20during.20.60EarlyPass.60/near/362825701):

- Add `default = ["internal"]` to the features in `src/tools/clippy/Cargo.toml`
- Remove the `#[inline]` attribute from `ty::tls::enter_context` if you want to break on it
- `x.py build src/tools/clippy`
- `LD_LIBRARY_PATH=build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib rust-gdb --args ./build/x86_64-unknown-linux-gnu/stage1-tools-bin/clippy-driver src/tools/clippy/tests/ui-internal/custom_ice_message.rs`
    - `break rustc_middle[9fd6dc49f5072439]::ty::context::tls::enter_context::`


This would be automatically tested, if we'd enable the `internal` feature when testing Clippy.

**Comment by Zoxc** [comment]
For https://github.com/rust-lang/rust/issues/112522, there's infinite stack recursion (collecting active queries -> running queries to get a query description -> triggers a cycle error -> collecting active queries -> ...). There's probably something similar happening with the Clippy ICE. I have a workaround in https://github.com/rust-lang/rust/pull/112603.

We should probably not OOM on infinite stack recursion though. The compiler uses stacker for unbounded segmented stacks.

**Comment by lukas-code** [comment]
Smaller reproducer:

```rust
fn main() {
    #[deny(while_true)]
    while true {}
}
```

```shell
rustc foo.rs -Ztreat-err-as-bug
```

**Comment by flip1995** [comment]
Thanks I'll add this test to the PR.

**Comment by flip1995** [comment]
When I add this test, I now get

```
running 1 tests
F

failures:

---- [ui] tests/ui/panics/panic-causes-oom-112708.rs stdout ----

error: Error: expected failure status (Some(101)) but received status None.
status: signal: 6 (SIGABRT) (core dumped)
...
```

because of the issue of the double panic that #112333 tried to fix:

```
query stack during panic:
thread panicked while processing panic. aborting.
```

Can I add some test flags, that would prevent that from happening?

**Comment by matthiaskrgr** [comment]
Is it because rustc "crashes" inside the test?

There are a couple of tests that crash intentionally, for example:

`tests/ui/treat-err-as-bug/delay_span_bug.rs`

```rust
// compile-flags: -Ztreat-err-as-bug
// failure-status: 101
// error-pattern: aborting due to `-Z treat-err-as-bug=1`
// error-pattern: [trigger_delay_span_bug] triggering a delay span bug
// normalize-stderr-test "note: .*\n\n" -> ""
// normalize-stderr-test "thread 'rustc' panicked.*\n" -> ""
// rustc-env:RUST_BACKTRACE=0

#![feature(rustc_attrs)]

#[rustc_error(delay_span_bug_from_inside_query)]
fn main() {}

```

**Comment by flip1995** [comment]
> Is it because rustc "crashes" inside the test?

No, that's not the issue. I have the same header as this file. The issue is, that rustc crashes, while crashing. So the `failure-status` becomes `None`, as opposed to `Some(101)`, which it should be for an ICE.

File: tests/ui/treat-err-as-bug/panic-causes-oom-112708.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// compile-flags: -Ztreat-err-as-bug
+// failure-status: 101
```
Comment (lukas-code) [inline tests/ui/treat-err-as-bug/panic-causes-oom-112708.rs:2]:
```suggestion
// dont-check-failure-status
```

Comment (flip1995) [reply tests/ui/treat-err-as-bug/panic-causes-oom-112708.rs]: Thanks, that worked!

**Comment by cjgillot** [comment]
Thanks!
@bors r+

**Comment by lqd** [comment]
is that spurious on this builder ?

@bors retry timeout on dist-x86_64-linux-alt

**Comment by flip1995** [comment]
> is that spurious on this builder ?

I don't think it is. It timed out because of the test I added here.... I really don't know enough about the query system to even start debugging this. So I'll need some help coming up with a proper fix.

**Comment by Zoxc** [comment]
It's possible the parallel compiler deadlocked during the test. https://github.com/rust-lang/rust/pull/112603 may also address that.

**Comment by flip1995** [comment]
@bors r-

This commit causes the rollup failures. So taking it out of the queue, until a hero comes around who can fix this bug without timeouts or OOMs.

```
..............................................................test [ui] tests/ui/treat-err-as-bug/panic-causes-oom-112708.rs has been running for a long time
```

2 times already.

**Comment by oli-obk** [comment]
I changed the test to only run with one thread for now, as I'm unable to reproduce the alt builder failure locally.

@bors p=5 rollup=never

File: tests/ui/treat-err-as-bug/panic-causes-oom-112708.rs [hunk 2]
```diff
@@ -1,4 +1,6 @@
-// compile-flags: -Ztreat-err-as-bug
+// This should run with more than one thread, too, but that can cause
+// deadlocks in query stack printing.
+// compile-flags: -Ztreat-err-as-bug -Zthreads=1
```
Comment (lqd) [inline tests/ui/treat-err-as-bug/panic-causes-oom-112708.rs:3]: @oli-obk we currently expect to always use `-Zthreads=1` with the parallel compiler

Comment (oli-obk) [reply tests/ui/treat-err-as-bug/panic-causes-oom-112708.rs:3]: oh... hmm... I guess I need to try to run that test in a loop then until it fails...

**Comment by oli-obk** [comment]
@bors r-

@bors retry

stuck again

**Comment by oli-obk** [comment]
Ok, I can repro locally now

**Comment by oli-obk** [comment]
> It's possible the parallel compiler deadlocked during the test. #112603 may also address that.

yes, that's it. There are two threads and they both are waiting for each other.

**Comment by oli-obk** [comment]
cc @Zoxc @compiler-errors https://github.com/rust-lang/rust/pull/112708/commits/4c9de4b842157c1f7f36bdf8c71a8e37dea19456 implements https://github.com/rust-lang/rust/pull/112603#issuecomment-1591794914

**Comment by flip1995** [comment]
Huh, why does it print a backtrace with `rustc-env:RUST_BACKTRACE=0`? It also doesn't print that backtrace for me locally.

**Comment by Zoxc** [comment]
Why does this still revert https://github.com/rust-lang/rust/pull/112333 if you're trying to remove the problematic query use?

**Comment by oli-obk** [comment]
I thought it was an orthogonal issue. I didn't actually check what that commit does 🙈

**Comment by oli-obk** [comment]
yea, seems resolved.

@flip1995 please just squash everything down to one commit

**Comment by flip1995** [comment]
Done. I kept it as 2 commits. One for adding the test and one for the fix. Something something TDD :grin:

**Pull request merged**